### ISSUE TITLE
remove KDTree Python implementation

### ISSUE THREAD

**Issue by rgommers** [scipy/scipy#8923]
It was decided a while ago [1] to remove the `KDTree` implementation in Python, and make `KDTree` an alias for `cKDTree`. 

TODO list for this:

1. make KDTree an alias for cKDTree
2. make KDTreeNode an alias for cKDTreeNode
3. keep all names working
4. only have KDTree and KDTreeNode in the documentation
5. keep kdtree.py for the test suite, but do not export its contents

[1] https://mail.python.org/pipermail/scipy-dev/2016-February/021266.html

**Comment by sturlamolden** [comment]
Also, we need ```scipy.spatial.KDTree``` for the test suite, as it is used to test ```cKDTree```. It should not be removed but moved to ```scipy.spatial.tests``` and kept as a fancy test class.

**Comment by sturlamolden** [comment]
As for complex types, we can use C++ templating and Cython fusedtypes to create a generic implementation that will cover all NumPy dtypes, and perhaps even custom types. It will add more bloat to SciPy, though. It is mostly a matter of scanning through the code and replacing ```double``` with ```DTYPE``` (or whatever), but it is not a small task.

**Comment by adeak** [comment]
It seems to me that as of https://github.com/scipy/scipy/pull/12852 the two classes are feature equivalent. Would it be possible to note this somewhere in the documentation (and maybe even point to either of the classes as a "preferred" version)?

I couldn't find any pointers a https://docs.scipy.org/doc/scipy/reference/spatial.html#module-scipy.spatial nor the individual doc pages, so I had to dive into github issues to understand what the difference is between the two classes (and my current understanding is "nothing but `cKDTree` might be barely slimmer").

**Comment by rgommers** [comment]
That sounds like a good idea to me

**Comment by adeak** [comment]
Thanks @rgommers, I think I can get around to open a PR for that sometime. Should `cKDTree` be the preferred alternative, on the basis of `KDTree` being a wrapper with the same API?

**Comment by sturlamolden** [comment]
The relationship between KDTree and cKDTree used to be the same as between Pickle and cPickle. It is a Python 2-ism. In the future we should kill KDTree and rename cKDTree.

**Comment by adeak** [comment]
@sturlamolden it seems to me that's the long-term goal of this issue :) But it would be nice if the current _status quo_ was documented so that the next person doesn't have to spend 5 minutes wondering which one to look at.

**Comment by sturlamolden** [comment]
Given that the name cKDTree is likely to be deprecated and they are otherwise similar, KDTree is just thin wrapper over cKDTree, I would suggest we recommend the name KDTree.

**Comment by rgommers** [comment]
Agreed, `KDTree` is the preferred name, and it would be good to document that.
